File: Trace2JSON/Trace2JSON/Parsers/CoreAnimationParser.m
//
//  CoreAnimationParser.m
//  Trace2JSON
//
//  Created by Itay Brenner on 6/21/20.
//  Copyright © 2020 Itaysoft. All rights reserved.
//

#import "CoreAnimationParser.h"
#import "InstrumentsPrivateHeaders.h"

@implementation CoreAnimationParser

- (void)readRow:(XRAnalysisCoreReadCursor *)cursor formatter:(XREngineeringTypeFormatter *)formatter result:(NSMutableArray *)result {
    while (XRAnalysisCoreReadCursorNext(cursor)) {
        BOOL resultObject = NO;
        XRAnalysisCoreValue *object = nil;
        resultObject = XRAnalysisCoreReadCursorGetValue(cursor, 0, &object);
        NSString *timestamp = resultObject ? [formatter stringForObjectValue:object] : @"";
        resultObject = XRAnalysisCoreReadCursorGetValue(cursor, 2, &object);
        double fps = resultObject ? [object.objectValue doubleValue] : 0;
        resultObject = XRAnalysisCoreReadCursorGetValue(cursor, 3, &object);
        double gpu = resultObject ? [object.objectValue doubleValue] : 0;
        
        NSDictionary *rowValue = @{
            @"timestamp": timestamp,
            @"fps": @((int)fps),
            @"gpuUsage": @((int)gpu)
        };
        [result addObject:rowValue];
    }
}

- (id) parseContext:(NSArray<XRContext *> *) contexts withRun:(XRRun *)run {
    NSMutableArray *result = [NSMutableArray new];
    
    XRContext *context = contexts[0];
    [context display];
    XRAnalysisCoreTableViewController *controller = Ivar(context.container, _tabularViewController);
    XRAnalysisCorePivotArray *array = controller._currentResponse.content.rows;
    XREngineeringTypeFormatter *formatter = IvarCast(array.source, _filter, XRAnalysisCoreTableQuery * const).fullTextSearchSpec.formatter;
    [array access:^(XRAnalysisCorePivotArrayAccessor *accessor) {
        [accessor readRowsStartingAt:0 dimension:0 block:^(XRAnalysisCoreReadCursor *cursor) {
            [self readRow:cursor formatter:formatter result:result];
        }];
    }];
    
    return result;
}

@end

File: Trace2JSON/Parsers/CoreAnimationParser.swift
//
//  CoreAnimationParser.swift
//  Trace2JSON
//
//  Created by Itay Brenner on 6/22/20.
//  Copyright © 2020 Itaysoft. All rights reserved.
//

import Foundation

class CoreAnimationParser: ParserProtocol {
    func parseContext(_ contexts: [XRContext], run: XRRun) -> InstrumentRun {
        let coreRun = CoreAnimationRun()
        coreRun.run = Int(run.runNumber())
        coreRun.runName = run.displayName()
        
        let context = contexts[0]
        context.display()
        
        guard let container = context.container() as? XRAnalysisCoreDetailViewController,
            let controller: XRAnalysisCoreTableViewController = RuntimeHacks.getIvar(instance: container, name: "_tabularViewController"),
            let array = controller._currentResponse()?.content.rows(),
            let filter: XRAnalysisCoreTableQuery = RuntimeHacks.getIvar(instance: array.source()!, name: "_filter"),
            let formatter = filter.fullTextSearchSpec()!.formatter() else {
                return coreRun
        }
        
        array.access({ (accessor) in
            accessor?.readRowsStarting(at: 0, dimension: 0, block: { (cursor) in
                self.readRow(cursor, formatter, coreRun)
            })
        })
        
        return coreRun
    }
    
    fileprivate func readRow(_ cursor: UnsafeMutableRawPointer?, _ formatter: XREngineeringTypeFormatter, _ run: CoreAnimationRun) {
        while(XRAnalysisCoreReadCursorNext(cursor)) {
            var objectFound: Int32 = 0
            var object: XRAnalysisCoreValue? = nil
            
            objectFound = XRAnalysisCoreReadCursorGetValue(cursor, 0, &object)
            let time = objectFound != 0 ? formatter.string(for: object) : ""
            objectFound = XRAnalysisCoreReadCursorGetValue(cursor, 2, &object)
            let fps = objectFound != 0 ? (object?.objectValue())! as! NSNumber : 0
            objectFound = XRAnalysisCoreReadCursorGetValue(cursor, 3, &object)
            let gpu = objectFound != 0 ? (object?.objectValue())! as! NSNumber : 0

            let row = CoreAnimationRow(timestamp: time ?? "",
                                    fps: fps.intValue,
                                    gpuUsage: gpu.intValue)
            run.result.append(row)
        }
    }
}
